ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionIndexingInfo.scala

```diff
@@ -0,0 +1,198 @@
+    val exerciseArguments = Map.newBuilder[NodeId, Array[Byte]]
+    val exerciseResults = Map.newBuilder[NodeId, Array[Byte]]
+
+    for ((nodeId, event) <- events) {
+      val eventId = EventId(transactionId, nodeId)
+      event match {
+        case create: Create =>
+          val (createArgument, createKeyValue) = translation.serialize(eventId, create)
+          createArgumentsByContract += ((create.coid, createArgument))
+          createArguments += ((nodeId, createArgument))
+          createKeyValue.foreach(key => createKeyValues += ((nodeId, key)))
+        case exercise: Exercise =>
+          val (exerciseArgument, exerciseResult) = translation.serialize(eventId, exercise)
+          exerciseArguments += ((nodeId, exerciseArgument))
+          exerciseResult.foreach(result => exerciseResults += ((nodeId, result)))
+        case _ => throw new UnexpectedNodeException(nodeId, transactionId)
+      }
+    }
+
+    for (DivulgedContract(contractId, contractInst) <- divulgedContracts) {
+      val serializedCreateArgument = translation.serialize(contractId, contractInst.arg)
+      createArgumentsByContract += ((contractId, serializedCreateArgument))
+    }
+
+    TransactionIndexingInfo.Serialized(
+      info = this,
+      createArgumentsByContract = createArgumentsByContract.result(),
+      createArguments = createArguments.result(),
+      createKeyValues = createKeyValues.result(),
+      exerciseArguments = exerciseArguments.result(),
+      exerciseResults = exerciseResults.result(),
+    )
+  }
+}
+
+object TransactionIndexingInfo {
+
+  private class Builder(blinding: BlindingInfo) {
+
+    private val events = Vector.newBuilder[(NodeId, Node)]
+    private val creates = Set.newBuilder[Create]
+    private val archives = Set.newBuilder[ContractId]
+    private val stakeholders = Map.newBuilder[NodeId, Set[Party]]
+    private val disclosure = Map.newBuilder[NodeId, Set[Party]]
+    private val visibility = Vector.newBuilder[(ContractId, Set[Party])]
+
+    private def addEventAndDisclosure(event: (NodeId, Node)): Unit = {
+      events += event
+      disclosure += ((event._1, blinding.disclosure(event._1)))
+    }
+
+    private def addCreation(create: Create): Unit =
+      creates += create
+
+    private def addArchival(contractId: ContractId): Unit =
+      archives += contractId
+
+    private def addDivulgence(contractId: ContractId): Unit =
+      for (divulgees <- blinding.divulgence.get(contractId)) {
+        visibility += ((contractId, divulgees))
+      }
+
+    private def addVisibility(contractId: ContractId, parties: Set[Party]): Unit =
+      visibility += ((contractId, parties))
+
+    private def addStakeholders(nodeId: NodeId, parties: Set[Party]): Unit =
+      stakeholders += ((nodeId, parties))
+
+    def add(event: (NodeId, Node)): Builder = {
+      event match {
+        case (nodeId, create: Create) =>
+          create.templateId.toString
```

Comment:
Probably the compiler will complain about this one, but just to make sure: it's unused, right?

Reply:
I don't think the compiler can make assumptions on whether that piece of code can be safely marked as unused, as it might perform side effects. But indeed it can be safely removed, thanks for the good catch. 🙂

Reply:
Addressed by https://github.com/digital-asset/daml/pull/8019/commits/13edba9143a2e1eb69e54d71716493f298b50d83